File: 23-06-2024_Print_Bracket_number.py
# from typing import List


class Solution:

    def bracketNumbers(self, str):
        ans = []  # List to store the bracket numbers
        st = []  # Stack to store the opening bracket numbers

        bal = 0  # Variable to keep track of current bracket number

        # Iterating over the string
        for i in range(len(str)):
            # If opening bracket is encountered
            if str[i] == '(':
                # Incrementing the bracket number
                bal += 1

                # Adding the bracket number to the result list and stack
                ans.append(bal)
                st.append(bal)
            # If closing bracket is encountered
            elif str[i] == ')':
                # Adding the top bracket number from the stack to the result list
                ans.append(st.pop())

        return ans  # Returning the result list with assigned bracket numbers

File: 23-06-2024_Print_Bracket_number.cpp
//Back-end complete function template for C++

class Solution {
  public:
    // Function to assign bracket numbers to each parenthesis in a string.
    vector<int> bracketNumbers(string str) {
        string temp = str;
        vector<int> ans; // vector to store the bracket numbers
        stack<int> st;   // stack to store the opening bracket numbers

        int bal = 0; // variable to keep track of current bracket number

        // iterating over the string
        for (int i = 0; i < temp.size(); i++) {
            // if opening bracket is encountered
            if (temp[i] == '(') {
                // incrementing the bracket number
                bal++;

                // adding the bracket number to the result vector and stack
                ans.push_back(bal);
                st.push(bal);
            }
            // if closing bracket is encountered
            else if (temp[i] == ')') {
                // adding the top bracket number from the stack to the result vector
                ans.push_back(st.top());
                st.pop();
            }
        }

        return ans; // returning the result vector with assigned bracket numbers
    }
};
